Add AF_UNIX support to HTTP connections

yetamrra opened the pull request:
Allow the URL for a device to be specified as unix://socket_name/... in
addition to the existing http:// URLs.  Also add a new socket_dir option
to the [options] section of airscan.conf.

When a unix:// URL is encountered, the sockdir_dir and socket_name
values are combined to form a full path to a local AF_UNIX socket.
Instead of attempting a TCP connection to the device, all HTTP requests
for that device are sent through the local socket.  The socket file
itself must be created by some external program that knows how to
communicate with the scanner, such as the Chrome OS ippusb_bridge that
forwards through an IPP-USB tunnel.

This also adds underscore to the list of accepted hostname characters in
http_parser.  Underscore is technically not legal in a hostname, but
it's extremely common in filenames and won't hurt parsing valid URLs.

Fixes alexpevzner/sane-airscan-unstable#11

yetamrra committed:
Add AF_UNIX support to HTTP connections

Allow the URL for a device to be specified as unix://socket_name/... in
addition to the existing http:// URLs.  Also add a new socket_dir option
to the [options] section of airscan.conf.

When a unix:// URL is encountered, the sockdir_dir and socket_name
values are combined to form a full path to a local AF_UNIX socket.
Instead of attempting a TCP connection to the device, all HTTP requests
for that device are sent through the local socket.  The socket file
itself must be created by some external program that knows how to
communicate with the scanner, such as the Chrome OS ippusb_bridge that
forwards through an IPP-USB tunnel.

The URL is parsed with the same parser as http:// URLs, so the socket
name is currently restricted to valid DNS hostname characters.

Fixes alexpevzner/sane-airscan-unstable#11

alexpevzner commented:
Wow! Big patch, indeed.

My few comments:
1. Let's move `http_uri_fix_scheme` logic into the `http_uri_new_relative`. http_uri_fix_host is very eSCL-specific trick, while prevention redirects from overriding scheme (and host too!) is applicable to the IPP-scan as well (and in theory to the WSD too, but WSD doesn't work over USB). So it is very generic, let's don't expose it to the protocol handlers
2. Can we avoid patching `http_parser.c`? RFC1035 doesn't allow underscores in domain names, but allow hyphens. Can we stay within this limitation?
3. Let's default `socket_dir` to '/var/run/`.
4. If we document `socked_dir` in the manual page, we should document handling of direct URL by `sane_open` as well :-)
5. You don't need to zero-fill memory, returned by `mem_new`, it does it automatically
6. You don't check that resulting path fits `sockaddr_un::sun_path` buffer.
7. In a case of `AF_UNIX`, `q->straddr` needs to be initialized into the something reasonable. Note, it is used only for logging

yetamrra commented:
For (2), we can change our socket names to avoid the _.  In general, it seems like if this is advertised as taking a file name, it will be a surprise for some people if it only accepts a subset of filenames.  Even the underscore doesn't fully fix this.  I was wondering if it would be better to special-case the unix scheme to accept anything that isn't / rather than just hacking in _ to the regular http scheme.  What do you think?

For the other points, I'll prep an updated PR that addresses them.

alexpevzner commented:
I will not seriously object against extending hostname interpretation for the special case of unix scheme, but I really want to avoid patching of `http_parser.c` (unless you will push this patch to the http_parser upstream :-)).

So if you want to go this way, it's better to handle it in the `http_uri_parse`

Alternatively, we can consider %-escaping (which looks ugly if used by humans, not internally by software). Please note, `http_uri_unescape_host` purposely handles only `%25` - it makes it more tolerant to IPv6 zone suffix, written with mistake (i.e., `[fe80::640e:2175:dfb5:5ddb%4]` instead of `[fe80::640e:2175:dfb5:5ddb%254]`)

alexpevzner commented:
BTW, I've noticed that you've replaced ippusbxd with ippusb_bridge, written in Rust. How much executable size did you save in comparison to Go ipp-usb?

yetamrra commented:
We wanted to switch to ipp-usb, but we couldn't get it small enough.  ippusb_bridge is ~625K, but it's not a completely fair comparison because it doesn't have all the features of ipp-usb.

alexpevzner commented:
Half of the ipp-usb size is the run-time type information, required for reflection and garbage collector, and half is the TLS implementation, which is not used, but still linked in, because has reference from the http package.

I've noticed that Go team currently puts a lot of effort into optimization of the resulting executables size, by avoiding linkage of unused types and by compacting run-time type information. I will not be surprised, if in 1-2 years ipp-usb size will be reduced to 1/4 of its current size.

yetamrra committed:
Document direct URL specification

The feature was added in commit d5ad808 ("airscan-zeroconf: Accept
direct device specification"), but wasn't documented at the time.

yetamrra commented:
1. It looks like `http_uri_new_relative` already had the logic to keep the original scheme, host, and port.  I've changed the call in `escl_scan_decode` to pass true for `path_only`.  That looks like the right move, assuming that we wouldn't want to follow a redirect to a new hostname based on the result of initiating a scan.  If there are cases where that isn't true, I can make it only restrict to the path for unix:// URLs.

2.  For now I've just reverted the change to http_parser.  We will change our socket paths to avoid the underscore.  I still think theoretically it might be better to accept other valid path characters, but I'd like to not make this change any bigger :)

3. Done by setting it in `conf_load`.

4. Added a second commit to document the direct URL passing.

5. Thanks, removed the memsets.

6. I was assuming that the truncated path would error out when `connect` is called, but you're right.  I added an explicit error return path.

7. I added a case to `ip_straddr_from_sockaddr_dport` in case anything else needs to convert the path to a string.  Do you think that's the right approach, or should I specifically just handle setting `q->straddr`?

alexpevzner commented:
1 - I'd prefer the following approach: lets `http_uri_new_relative` assume `path_only == true` if `scheme` is `HTTP_SCHEME_UNIX`. See comments to https://github.com/alexpevzner/sane-airscan-unstable/commit/3bbdef921a885dfeff21015c71a4ded967db777c for details

2  - OK :-)

7  - I could accept it, but at this case buffer in the `ip_straddr` needs to be expanded.

yetamrra commented:
I made the ip_straddr buffer 109 because 108 is the length of sun_path on Linux.  Other OSs that I checked have the same or shorter.

Any other concerns with this patch?

alexpevzner commented:
Hi Benjamin,

sorry for delay, I was very busy these days. I've briefly reviewed your patch, at the first glance everything looks OK. Today I'll try to find a time to review it more detailed, and if everything will be still OK, I'll accept it. If there will be only minor notes from my side, I'll fix it by myself.

alexpevzner commented:
Hi,

I just want to inform you, that this patch has propagated up to the freshly released 0.99.17 version of sane-airscan

yetamrra commented:
> Hi,
> 
> I just want to inform you, that this patch has propagated up to the freshly released 0.99.17 version of sane-airscan

Great news.  Thanks!
